cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/status.h"
+
+namespace arrow {
+
+namespace io {
+
+class InputStream;
+class OutputStream;
+
+}  // namespace io
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// secondes since the epoch.
+using TimePoint = std::chrono::sys_time<std::chrono::nanoseconds>;
+
+enum class FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a folder and contents.
+  virtual Status DeleteFolder(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  virtual Status DeleteFiles(const std::vector<string>& path) = 0;
```

Comment:
Is there a concrete use-case for batch delete?  I would imagine most file systems don't support transactional type deletes, so it might pay to expose this at a higher level instead of the core filesystem

Reply:
I suspect something like deleting a Parquet partition with a large number of files?
The intention is not to allow transactional deletes but rather to avoid a roundtrip per each individual delete (which may hurt a lot if accessing a remote filesystem with 100+ms latencies).

Reply:
Yes, certainly, some file-systems will provide shortcuts for bulk operations or various types, and they will be *very important* in many use cases. In general, it would be exceptionally useful if file-systems provide glob and directory-tree-based operations.

Reply:
Do we need globbing at the filesystem level or can we give the client a complete list of directory contents and do the filtering on the client side?

Reply:
Might want to make the default implementation of this call `DeleteFile` on each path. Are there any filesystems that do support multiple-deletes in a single function call / RPC?

Reply:
I can't say what you "need", but listing all the contents of some S3 bucket seems like a really bad idea. Glob functionality is something that users will find very convenient, and would best be done in one place only (ref https://github.com/martindurant/filesystem_spec/blob/master/fsspec/spec.py#L375 ).

Reply:
>  but listing all the contents of some S3 bucket seems like a really bad idea

Does the S3 API support generic globbing?

Reply:
> Glob functionality is something that users will find very convenient, and would best be done in one place only 

Right, but should it be implemented by each filesystem (which implies it can be done on the server side) or should it be an agnostic facility in the client? Your reference to the `fsspec` code seems to suggest the latter.

Reply:
A generic version should be provided, and some implementations (like local) might be able to provide their own.

Reply:
My question was: is it _useful_ to have a filesystem-specific implementation, and if yes, in which situations?

Reply:
"yes" - e.g., S3, where deleting one file takes ~>100ms, and bulk deleting many files also takes the same time. By time, I mean how long it takes for a response from the server - they may take longer to disappear from file listings.

Reply:
We're not understanding each other here.

i'm asking about server-side _globbing_, not about batch deletion.

Reply:
If the server can do glob, rather than having to issue many list-dir commands and filter ourselves, this would apply to that too. I'm not sure that any of the services you are thinking about here do - the only one I can think of is SSH.

Reply:
Note that recursively listing, in the current proposal, is possible with `GetTargetStats(const Selector& select, ...)`.

Reply:
I do not think that Arrow should be concerned with deleting upstream files. At least in this iteration.

Reply:
Pragmatically, deleting files can be useful for testing :-)

Reply:
@fsaintjacques we need to be able to delete files to roll back incomplete writes